Additional tests for ufunc(Series)

TomAugspurger opened the pull request:
This adds a set of tests for ufuncs on Series. The goal is to
establish the correct behavior prior to implementing `Series.__array_ufunc__`.

There are two kinds of xfails right now

1. Series[Sparse] fails because `Series.__array_ufunc__` doesn't
    yet dispatch to `Series.array.__array_ufunc__`
2. `ufunc(series, series)` when the two series are unaligned. It's
    been determined that these should align, but isn't currently
    implemented.

LMK if there's anything I can do to make these tests clearer. The heavy parameterization is a bit unfortunate, but I think the best option.

xref https://github.com/pandas-dev/pandas/pull/23293

TomAugspurger committed:
TST: Additional tests for Series ufuncs

This adds a set of tests for ufuncs on Series. The goal is to
establish the correct behavior prior to implementing `Series.__array_ufunc__`.

There are two kinds of xfails right now

1. Series[Sparse] fails because `Series.__array_ufunc__` doesn't
   yet dispatch to `Series.array.__array_ufunc__`
2. `ufunc(series, series)` when the two series are unaligned. It's
   been determined that these should align, but isn't currently
   implemented.

TomAugspurger committed:
fixup release note

TomAugspurger commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,152 @@
+import pandas.util.testing as tm
+
+UNARY_UFUNCS = [np.positive, np.floor, np.exp]
+BINARY_UFUNCS = [np.add, np.logaddexp]  # -> dunder op
```
whoops, this comment is supposed to be for `np.add`.

jbrockmendel replied:
Any reason not to go all out and use the full list of ufuncs here?

TomAugspurger replied:
Three reasons

1. I couldn't find a fixture with all the ufuncs (I thought we had one, but apparently not)
2. I have to treat binary input and binary output separately, so I need separate tests anyway
3. I'm mostly concerned with the dispatching logic. I don't think the marginal value of testing `np.negative` is high, if we're already testing `np.positive`

TomAugspurger commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,152 @@
+        a2 = pd.SparseArray(a2, dtype=pd.SparseDtype('int', 0))
+
+    name = "name"
+    # TODO: verify name when the differ? Take the first? Drop?
```
I think `Series.add(series)` is informative here. That decides to drop the name. I'll likely add a new test exercising just that.

jbrockmendel replied:
There are some arithmetic tests that do this, might be worth making into a fixture.  grepping for "Venkman" should find it.

TomAugspurger commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,152 @@
+@pytest.mark.parametrize("sparse", SPARSE, ids=SPARSE_IDS)
+@pytest.mark.parametrize("shuffle", SHUFFLE)
+@pytest.mark.parametrize("box_other", [True, False])
+def test_binary_ufunc(ufunc, sparse, shuffle, box_other,
```
This test should probably be parameterized to do both `ufunc(a, b)` and `ufuc(b, a)`. This matters for `ufunc(Series, Array)`. I think it's preferable for the `Array.__array_ufunc__` to return `NotImplemented` so that `Series.__array_ufunc__` is called. Does that sound good? It's what xarray does with DataArray and ndarray.

jreback replied:
yes this sounds good

jbrockmendel commented:
Should DTA[tz] be included along with Sparse?

TomAugspurger commented:
> Should DTA[tz] be included along with Sparse?

IMO, the intention of these tests is to verify the correctness of the future `Series.__array_ufunc__`. I parametrize over `Series[int]` and `Series[Sparse[int]]` to verify the correctness of the dispatch to array logic. In *theory* the dispatch logic will handle all arrays the same. I'll add a small sanity test, separate from the rest, since the set of ufuncs valid for datetime64 is different.

jbrockmendel commented:
OK, many of the DTA[tz] ufuncs are currently broken, so its OK by me if you want to leave them out.

TomAugspurger committed:
fixups

TomAugspurger committed:
remove stale comment

TomAugspurger committed:
xfail ufunc(series, index)

TomAugspurger commented:
0b1e745 adds another xfail for `ufunc(Index, Series)`, e.g. `np.add(index, series)`. Right now that returns an Index

```python
In [1]: import pandas as pd; import numpy as np

In [2]: a = pd.Series([1, 2])

In [3]: b = pd.Index([1, 2])

In [4]: np.add(b, a)
Out[4]: Int64Index([2, 4], dtype='int64')
```

I think that's a bug. It's inconsistent with the binop

```python
In [1]: import pandas as pd; import numpy as np

In [2]: a = pd.Series([1, 2])

In [3]: b = pd.Index([1, 2])

In [4]: np.add(b, a)
Out[4]: Int64Index([2, 4], dtype='int64')
```

So the rule is that `Index.__array_ufunc__` should check the `types` for Series. If any are found then it should return NotImplemented, so that `Series.__array_ufunc__` can take over.

TomAugspurger committed:
32-bit compat

(cherry picked from commit ba5cb55e511ea484bf4a63d4c489767e5bed87a2)

TomAugspurger committed:
fixup

TomAugspurger commented:
https://github.com/pandas-dev/pandas/pull/26951/commits/775c2efde0f84dc7f1a2c214bc6d8a0bd21f1d4d fixes an issue with one of the tests. The test wasn't correct for `ufunc(a, b)` when we shuffle b (to force alignment). It was passing earlier, since we xfail `shuffle=True` for now.

I think this PR should be ready to go. I'd like to merge Joris' `array_ufunc` branch on top of it to actually start fixing the xfails.

jreback commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,185 @@
+@pytest.mark.parametrize("ufunc", BINARY_UFUNCS)
+@pytest.mark.parametrize("sparse", SPARSE, ids=SPARSE_IDS)
+@pytest.mark.parametrize("shuffle", SHUFFLE)
+@pytest.mark.parametrize("box_other", ['series', 'index', 'raw'])
```
can you raw -> ndarray

jreback commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,185 @@
+def test_binary_ufunc(ufunc, sparse, shuffle, box_other,
+                      flip,
+                      arrays_for_binary_ufunc):
+    # Check the invariant that
```
can you give a little expl of what the parametizations do if not obvious, e.g. flip & shuffle actually are not immediately obvious

TomAugspurger replied:
I may split those out to separate tests. It'll be a bit more code, but *much* clearer.

jreback commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,185 @@
+    # Check the invariant that
+    #   ufunc(Series(a), Series(b)) == Series(ufunc(a, b))
+    #   with alignment.
+    a1, a2 = arrays_for_binary_ufunc
```
can you call these: left_array, right_array

I know its a bit longer, but more readable IMHO

jreback commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,185 @@
+        a2 = pd.SparseArray(a2, dtype=pd.SparseDtype('int', 0))
+
+    name = "name"
+    s1 = pd.Series(a1, name=name)
```
left_series and right_series

jreback commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,185 @@
+            a2 = a2.take(idx)
+
+    a, b = s1, s2
+    c, d = a1, a2
```
I find this very confusing, can you use the same terms and not use a, b, c, d?

jreback commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,185 @@
+    series = pd.Series(array, name="name")
+
+    a, b = series, other
+    c, d = array, other
```
same as above if you can make this more clear

jreback commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,185 @@
+    tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize("ufunc", [np.divmod])  # any others?
```
np.modf (what you just added)

TomAugspurger replied:
modf is a unary function with two outputs. This test is for binary functions with one output.

jreback commented on pandas/tests/series/test_ufunc.py:
```diff
@@ -0,0 +1,185 @@
+        array = pd.SparseArray(array)
+
+    series = pd.Series(array, name="name")
+    result = np.modf(series)
```
ok I c you are doing this here

jreback reviewed (changes requested):
seem reasonble, just some readability comments

jreback commented:
do we have an issue for this? (or maybe just xref the actual ufunc issue at the top of the PR)

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into series-array-ufunc

TomAugspurger committed:
more

TomAugspurger committed:
lint

TomAugspurger commented:
Changes

* split the binop ufunc test into three: with array, with index, with Series. Simplified things quite a bit I think
* Cleaned up the names. Hopefully `array_args` and `series_args` make things much clearer that `a, b, c, d`.

jreback commented:
lgtm.

jreback commented:
thanks @TomAugspurger
